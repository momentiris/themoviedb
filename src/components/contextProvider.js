import React, { createContext, useReducer } from 'react'

export const initialState = {
  loading: false,
  query: '',
  error: '',
  fetching: false,
  searchResult: {
    movies: [],
    movie: {},
    credits: [],
  },
}

const types = {
  LOADING: 'loading',
  QUERY: 'query',
  ERROR: 'error',
  FETCHING: 'fetching',
  SEARCH_RESULT_MOVIES: 'search result movies',
  SEARCH_RESULT_MOVIE: 'search result movie',
  SEARCH_RESULT_CREDITS: 'search result credits',
}

export const reducer = (state, action) => {
  switch (action.type) {
    case types.LOADING:
      return { ...state, loading: !state.loading }
    case types.QUERY:
      return { ...state, query: action.payload }
    case types.SEARCH_RESULT_MOVIES:
      return {
        ...state,
        searchResult: { ...state.searchResult, movies: action.payload },
      }
    case types.SEARCH_RESULT_MOVIE:
      return {
        ...state,
        searchResult: { ...state.searchResult, movie: action.payload },
      }
    case types.SEARCH_RESULT_CREDITS:
      return {
        ...state,
        searchResult: { ...state.searchResult, credits: action.payload },
      }
    case types.ERROR:
      return { ...state, error: action.payload }
    default:
      return
  }
}

const Context = createContext(initialState)

export { types, Context }

export default ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState)

  return (
    <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>
  )
}
