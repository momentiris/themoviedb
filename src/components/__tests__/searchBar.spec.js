import React from 'react'
import SearchBar from '../searchBar'
import { render, fireEvent } from 'react-testing-library'
import { Context, types } from '../contextProvider'

describe('components/searchBar', () => {
  it('renders component', () => {
    const { container } = render(<SearchBar />)

    expect(container).toMatchSnapshot()
  })
})

test('dispatches new search value', async () => {
  const dispatch = jest.fn()
  const { container } = render(
    <Context.Provider value={{ dispatch }}>
      <SearchBar />
    </Context.Provider>
  )

  const searchbar = container.querySelector('input')
  fireEvent.change(searchbar, { target: { value: 'Batman' } })
  fireEvent.click(container.querySelector('button'))

  expect(dispatch).toHaveBeenCalledWith({
    type: types.QUERY,
    payload: 'Batman',
  })
})
