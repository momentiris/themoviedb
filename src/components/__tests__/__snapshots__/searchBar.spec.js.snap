// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/searchBar renders component 1`] = `
<div>
  <div
    class="sc-bdVaJa cdlyjo"
  >
    <input
      class="sc-bwzfXH fodZgz"
      placeholder="Search for something... or don't, I don't care."
    />
    <button
      class="sc-htpNat fQXDrD"
    >
      Search
    </button>
  </div>
</div>
`;
